testify/test_rerunner.py

```diff
@@ -32,23 +42,29 @@ def discover(self):
         else:
             tests = open(self.filename)
 
-        with tests as tests:
-            tests = tests.read()
+        test = tests.readline().rstrip()
+
+        if not test:
+            return
 
-        if tests.startswith('{'):
+        if test.startswith('{'):
             constructor = Test.from_json
         else:
             constructor = Test.from_name
 
-        tests = [
-            constructor(test)
-            for test in tests.splitlines()
-            if test  # Skip blank lines
-        ]
+        # Put back the test name we peeked at to choose a constructor
+        tests = chain(
+            [constructor(test)],
+            (
+                constructor(test.rstrip())
+                for test in readlines(tests)
+                if test  # Skip blank lines
+            ),
+        )
 
         for class_path, tests in groupby(tests, lambda test: test[:2]):
-            methods = [test.method for test in tests]
             module, cls = class_path
+            methods = (test.method for test in tests)
```

Comment:
given it's already gone through `groupby` making this into a generator is likely to *increase* the memory (as it now needs to keep the closure scope alive), and slower because generators